zip-0244.rst

```diff
@@ -1,8 +1,603 @@
+
+1: ``transparent_scripts_digest``
+`````````````````````````````````
+A BLAKE2b-256 hash of the field encoding of the Bitcoin script associated 
+with each transparent input belonging to the transaction.
+
+The personalization field of this hash is set to::
+
+  "ZTxAuthTransHash"
+
+2: ``sprout_auth_digest``
+```````````````````````````
+A BLAKE2b-256 hash of the field encoding of the zkproof values of each
+``JSDescription`` belonging to the transaction.
+
+   * 2a. ``zkproofs``         (field encoding bytes)
+   * 2b. ``joinsplit_pubkey``    
+   * 2b. ``joinsplit_sig``     
+
+The personalization field of this hash is set to::
+
+  "ZTxAuthSprouHash"
+
+3: ``sapling_auth_digest``
+```````````````````````````
+A BLAKE2b-256 hash of the field encoding of the Sapling zkproof values
+of each Sapling spend description and output description belonging to 
+the transaction, followed by the field encoding of the binding signature ::
+
+   * 3a. ``spend_zkproofs``  (field encoding bytes)
+   * 3b. ``output_zkproofs`` (field encoding bytes)
+   * 3c. ``binding_sig``     
+
+The personalization field of this hash is set to::
+
+  "ZTxAuthSapliHash"
+
+--------------------
+Block Header Changes
+--------------------
+
+The nonmalleable transaction identifier specified by this ZIP will be used
+in the place of the current malleable transaction identifier within the
+Merkel tree committed to by the ``hashMerkleRoot`` value. However, this 
+change now means that ``hashMerkleRoot`` is not sufficient to fully commit
+to the transaction data, including witnesses, that appear within the block.
+
+As a consequence, we now need to add a new commitment to the block header.
+This commitment will be the root of a Merkle tree that has parallel structure
+to the tree committed to by ``hashMerkleRoot`` (a path through this merkle
+tree to a transaction identifies the same transaction as that path reaches
+in the tree rooted at ``hashMerkleRoot``) but where the leaves are hashes
+produced according to the :ref:`Authorizing Data Commitment<Authorizing Data Commitment>` 
+specification.
+
+This new commitment is named ``hashAuthDataRoot`` and is the root of a left-dense
+binary merkle tree of transaction authorizing data commitments. Empty internal nodes 
+and leaves in the merkle tree (nodes without children) have the "null" hash value 
+``[0u8; 32]``. Hashes in this tree are BLAKE2b-256 hashes personalized by the string
+``"ZcashAuthDatHash"``.
+
+Changing the block header format to allow space for an additional 
+commitment is somewhat invasive. Instead, the name and meaning of the 
+``hashLightClientRoot`` field is changed.
+
+``hashLightClientRoot`` is renamed to ``hashBlockCommitments``. The value
+of this hash is the BLAKE2b-256 hash personalized by the string ``"ZcashBlockCommit"``
+of the following elements::
+
+   * ``hashLightClientRoot`` as described in ZIP 221 [#zip-0221]_
+   * ``hashAuthDataRoot`` 
+   * ``terminator`` [0u8;32]
```

Comment:
Are we going to increment the version field in the block header because we changed the meaning of this field?
Would a version change break other software which ignores the `hashLightClientRoot` field?

Reply:
This change is intended to be deployed as affecting the version 5 transaction format, which will likely also include TZEs (https://zips.z.cash/zip-0222) and the Orchard (https://zips.z.cash/zip-0224) protocol. The definition of V5 transactions will be laid out completely in the main protocol spec and a separate ZIP defining all the changes for the network upgrade in which it activates. The current thinking is that both the current V4 transaction format, with the old txid and signature algorithm, and the new transaction format will be supported on the network for some period of time, but there hasn't been a lot of effort put toward that decision as of yet.

Reply:
I think you're saying that the block version (the `nVersion` field in the block header) won't change, but the transaction version will.

Or maybe you're saying that the final decision will be made in the network upgrade ZIP that includes this ZIP?

If we're going to wait until the network upgrade ZIP to make the block version decision, let's make a recommendation in this ZIP. I suggest that we increment the block version, because we changed the meaning of a block header field in an incompatible way.

Reply:
We are not proposing to change the block version. The rationale is similar to when we did not change it for Blossom, even though we changed the semantics of this field. An additional argument is that, because of the Blossom upgrade, we now know that existing mining software does the right thing when the meaning of this field is changed but the block version is not.

The design with `terminator` means that in future we will not need to change the top-level definition (or name) of this field again in order to commit to additional information.

Reply:
@teor2345 
> Would a version change break other software which ignores the `hashLightClientRoot` field?

Yes. For one, the Zcash Stratum protocol (used AFAIK by mining pools to communicate with miners) requires that the block header version is 4, to allow for block header version changes to make arbitrary changes to the block header.

If we were going to change the block version, we should have done so in Heartwood when FlyClient was deployed. The reason we did not is partly because the FlyClient ZIP authors chose to write the ZIP such that this wouldn't happen, and partly because changing the block header version has a significant engineering overhead on the mining ecosystem (in the consensus rules it's a free non-negative integer field, so we can't rely on it for consensus checks without doing the sign-bit trick we did for transactions, and doing _that_ means _mining pools_ writing new code in their internal stacks to support the new header format). This cost wasn't justifiable for Heartwood, and I doubt it's justifiable for this change either, given that this field has only ever been used by light clients (and to my knowledge, is still unused by light clients - no one using `lightwalletd` is consuming FlyClient proofs, for instance, because it lacks support). What we _should_ have done is made the change proposed here in either the Sapling or Heartwood upgrades, but it was unfortunately missed.